pkg/provisioner/ironic/ironic.go

```diff
@@ -951,7 +951,7 @@ func (p *ironicProvisioner) Deprovision(deleteIt bool) (result provisioner.Resul
 
 	// FIXME(dhellmann): handle CleanFailed?
 
-	case nodes.Manageable:
+	case nodes.Manageable, string(nodes.Enroll), nodes.Verifying:
```

Comment:
This cast strikes me as odd, looks like the rest of the types are declared incorrectly.  I filed https://github.com/gophercloud/gophercloud/issues/1557 to fix this.

Reply:
Thanks, yeah i copied it from earlier in the file where @dhellmann already noted it's probably a gophercloud bug

Reply:
Makes sense, it's just the first time I noticed it.  The workaround is fine for now, will fix in gophercloud when I have a chance.

Reply:
Hi there,

I saw this comment linked from https://github.com/gophercloud/gophercloud/issues/1557.

I think there are two parts to this:

1. The shortcut for declaring consecutive constants of a specific type does not work (https://github.com/gophercloud/gophercloud/blob/master/openstack/baremetal/v1/nodes/requests.go#L21): This threw me for a loop, too, especially since `nodes.ListOpts.ProvisionState` will gladly accept strings and `ProvisionState` types. I'm wondering if this was some kind of change in the core language or something I've somehow just been able to avoid).

2. The case statement on line 954 here. This is a little confusing, so do let me know if I'm explaining poorly:

In Gophercloud, creating a custom type around a basic type (int, string, etc) is used as a form of _input_ validation. You can think of everything in the `requests.go` file as all input stuff.

`results.go` contains "output". We never use custom types here (the exception being for structs, but that's only to describe a set of struct fields). In the [`nodes.Node`](https://github.com/gophercloud/gophercloud/blob/master/openstack/baremetal/v1/nodes/results.go#L68) result struct, you can see that it has a `ProvisionState` field, which is what is being compared in the switch statement on line 895. 

The individual case statements (lines 897,  927, 934, etc) should technically be strings (`"error"`, `"available"`, `"deleting"`, etc). It is only by coincidence that the various `const`'s in `requests.go` were created as they were that these case statements work. 

So it's all really one big coincidence: A bunch of `const`'s were inadvertently declared as strings and someone saw a bunch of keyword constants declared so they decided to use them in an area which just so happened to want strings being used. Except for that _one_ `nodes.Enroll` const which shined light on all of this 😄 

It's, of course, a best practice to try to minimize the amount of times arbitrary strings are declared, so keyword constants are best used. If you wanted to continue using the `nodes.ProvisionState` suite of keywords, the best thing to do is to modify `requests.go` to explicitly declare all of them as a `ProvisionState` and then modify everywhere you're doing a string comparison to cast the const as a string.

Again, let me know if this makes sense.

Reply:
Thanks @jtopjian for replying here!

Yes that does make sense, I think the confusion stems from familiarity with the way the python-*client OpenStack libraries commonly work, where those do define the API surface for the language bindings to each service, and generally that does include some keyword constants to reduce duplication.

I can see you're only using these for input schema validation, and we know that in this case the input/output is the same so it'll probably work fine, but perhaps we should just maintain some constants locally to this repo instead, for the avoidance to future confusion.

Reply:
The string values are part of Ironic's API, so I would expect the client library to declare the constants even if they're just string objects and not a special type. @jtopjian, would you object if we added a set of response constants to gophercloud?

Reply:
The rest of gophercloud uses this pattern of responses being strings, but requests using constants. If the API returns something unexpected, I think you run into problems if the response is unexpected.

I did open a PR already to fix the types (https://github.com/gophercloud/gophercloud/pull/1558). After that, the only cast that's needed would be the beginning of the case statement, and then you can use the request types as the cases you want to handle.

```go
switch state := nodes.ProvisionState(ironicNode.ProvisionState); state {
```

Reply:
I wouldn't object to a set of _string_ constants, no. The important thing is to make sure that the fields of the Nodes result struct is composed of basic types. (side note: there are probably two or three places in Gophercloud where there's a "special" int or string type in results - these just need cleaned up some day)

Since we're now dealing with two sets of constants, you'll need to create different names for the set of strings.

Reply:
> If the API returns something unexpected, I think you run into problems in the unmarshalling of the response.

This is correct.

But if the constant is left as a plain string, then it can safely be used for comparisons/conditionals/switches as a convenience. This would be complementary to unmarshalling and not affect it at all.

Reply:
> I wouldn't object to a set of _string_ constants, no. The important thing is to make sure that the fields of the Nodes result struct is composed of basic types. (side note: there are probably two or three places in Gophercloud where there's a "special" int or string type in results - these just need cleaned up some day)
> 
> Since we're now dealing with two sets of constants, you'll need to create different names for the set of strings.

Sure, simple strings are fine. The main point from my perspective is to have the client provide some guidance about what values are likely in that response field.